externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp

```diff
@@ -84,6 +84,254 @@ OpFoldResult TMTensor::getDim(OpBuilder &builder, Location loc, Value v,
+  weightDynSizes[weightDynSizes.size() - 1] =
+      keyDynSizes[keyDynSizes.size() - 1];
+  Value weight = b.create<memref::AllocOp>(loc, weightType, weightDynSizes);
+  matmul(b, loc, query, queryDynSizes, key, keyDynSizes, weight, weightDynSizes,
+         true);
+
+  // weight = softmax(weight)
+  Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value dim = weightDynSizes[weightDynSizes.size() - 1];
+  Value scaleFactor = b.create<math::SqrtOp>(
+      loc, b.create<arith::UIToFPOp>(
+               loc, elementType,
+               b.create<arith::IndexCastUIOp>(loc, b.getI32Type(),
+                                              queryDynSizes[queryRank - 1])));
+  // calculate max(weight)
+  Value init = b.create<memref::LoadOp>(
+      loc, weight, SmallVector<Value>(weightDynSizes.size(), zero));
+  Value globalMax =
+      b.create<scf::ParallelOp>(
+           loc, SmallVector<Value>(queryRank, zero), weightDynSizes,
+           SmallVector<Value>(queryRank, one), init,
+           [&](OpBuilder &b, Location loc, ValueRange localIVs,
+               ValueRange accs) {
+             b.create<scf::ReduceOp>(
+                 loc, init,
+                 [&](OpBuilder &b, Location loc, Value elem, Value acc) {
+                   Value x = b.create<memref::LoadOp>(loc, weight, localIVs);
+                   Value max = b.create<arith::MaxFOp>(loc, x, acc);
+                   b.create<scf::ReduceReturnOp>(loc, max);
+                 });
+           })
+          .getResult(0);
+  // weight = (weight - max(weight)) / math.sqrt(querySizes[-1])
+  b.create<scf::ParallelOp>(
+      loc, SmallVector<Value>(queryRank, zero), weightDynSizes,
+      SmallVector<Value>(queryRank, one),
+      [&](OpBuilder &b, Location loc, ValueRange localIVs) {
+        Value x = b.create<memref::LoadOp>(loc, weight, localIVs);
+        x = b.create<arith::SubFOp>(loc, x, globalMax);
+        x = b.create<arith::DivFOp>(loc, x, scaleFactor);
+        b.create<memref::StoreOp>(loc, x, weight, localIVs);
+        b.create<scf::YieldOp>(loc);
+      });
+  // calculate exp(weight)
+  b.create<scf::ParallelOp>(
+      loc, zero, dim, one,
+      [&](OpBuilder &b, Location loc, ValueRange localIVs) {
+        Value outerDim = localIVs[0];
+        SmallVector<Value> min(queryRank - 1, zero),
+            max(weightDynSizes.begin(), weightDynSizes.end() - 1),
+            steps(queryRank - 1, one);
+        b.create<scf::ParallelOp>(
+            loc, min, max, steps,
+            [&](OpBuilder &b, Location loc, ValueRange localIVs) {
+              SmallVector<Value> coords(localIVs);
+              coords.push_back(outerDim);
+              Value x = b.create<memref::LoadOp>(loc, weight, coords);
+              x = b.create<math::ExpOp>(loc, x);
+              b.create<memref::StoreOp>(loc, x, weight, coords);
+              b.create<scf::YieldOp>(loc);
+            });
+      });
```

Comment:
What is the reason for using two `ParallelOp` here? I would've expected `exp(weight)` to be a single `ParallelOp` over the `weight` tensor.